tpu-provisioner/test/integration/controller/deletion_controller_test.go

```diff
@@ -0,0 +1,216 @@
+			node: makeNodeWithLabels("node3", map[string]string{
+				cloud.LabelNodepoolManager: cloud.LabelNodepoolManagerTPUPodinator,
+				cloud.GKENodePoolNameLabel: "test-nodepool-1",
+				cloud.LabelJobSetName:      defaultJobSetName,
+			}),
+			jobSet:               makeJobSet(defaultJobSetName),
+			jobSetStatus:         jobSetCompletedStatus,
+			wantNodePoolDeletion: true,
+		}),
+		Entry("node managed by provisioner and jobset is failed, node pool should be deleted", &testCase{
+			node: makeNodeWithLabels("node4", map[string]string{
+				cloud.LabelNodepoolManager: cloud.LabelNodepoolManagerTPUPodinator,
+				cloud.GKENodePoolNameLabel: "test-nodepool-1",
+				cloud.LabelJobSetName:      defaultJobSetName,
+			}),
+			jobSet:               makeJobSet(defaultJobSetName),
+			jobSetStatus:         jobSetFailedStatus,
+			wantNodePoolDeletion: true,
+		}),
+	)
+})
+
+// assertNodePoolDeletionTriggered validates that the given pod did trigger a node pool
+// deletion by checking that we eventually see the provider make a node pool deletion attempt.
+func assertNodePoolDeletionTriggered(node *corev1.Node) {
+	Eventually(func() bool {
+		_, deleted := provider.getDeleted(node.Name)
+		return deleted
+	}, timeout, interval).Should(BeTrue())
+}
+
+// assertNodePoolDeletionNotTriggered validates that the given pod did not trigger a node pool
+// deletion by checking that we consistently see the provider has made no node pool deletion attempts
+// for thet given pod.
+func assertNodePoolDeletionNotTriggered(node *corev1.Node) {
+	Consistently(func() bool {
+		_, deleted := provider.getDeleted(node.Name)
+		return deleted
+	}, timeout, interval).Should(BeFalse())
+}
+
+// makeNodeWithLabels returns a Node object with the given name and labels.
+func makeNodeWithLabels(name string, labels map[string]string) *corev1.Node {
+	return &corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   name,
+			Labels: labels,
+		},
+		Spec: corev1.NodeSpec{},
+	}
+}
+
+// makeJobSet returns an empty JobSet with the given name and namespace.
+func makeJobSet(name string) *jobset.JobSet {
+	return &jobset.JobSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: jobset.JobSetSpec{
+			ReplicatedJobs: []jobset.ReplicatedJob{},
+			Network:        &jobset.Network{},
+		},
+	}
+}
+
+func updateJobSetStatus(ctx context.Context, k8sClient client.Client, js *jobset.JobSet, status *jobset.JobSetStatus) {
+	// Fetch latest JobSet resource version so we don't get an operation conflict.
+	var createdJS jobset.JobSet
+	Eventually(func() error {
+		return k8sClient.Get(ctx, types.NamespacedName{Name: js.Name, Namespace: js.Namespace}, &createdJS)
+	}, timeout, interval).Should(Succeed())
+
+	By("Updating JobSet status")
+	createdJS.Status = *status
+	Eventually(func() error {
+		return k8sClient.Status().Update(ctx, &createdJS)
+	}, timeout, interval).Should(Succeed())
+}
+
+func deleteNode(ctx context.Context, k8sClient client.Client, node *corev1.Node) error {
```

Comment:
nit: prefer name like `ensureNodeDeleted()`

Reply:
Hmm, it isn't really ensuring it though, it just makes one attempt and returns the result, if any. Why do you think "ensure" is clearer?

Reply:
nvm the suggestion, I was thinking b/c it doesnt alway call .Delete()